# Table of Contents:

### 01. Importing Libraries
### 02. Importing Data
### 03. Creating a "price_label" column
### 04. Creating if statements with the loc() function
### 05. Creating a "busiest_day" column
### 06. Creating new "busiest_days" column
### 07. Creating a "busiest_period_of_day" column
### 08. Exporting Data 


# 01. Importing Libraries

In [6]:
#Import the pandas, NumPy, and os libraries.
import pandas as pd
import numpy as np
import os

# 02. Importing Data

In [8]:
#Creating a path
path = r'C:\Users\canam\Achievement 4\Instacart Basket Analysis'

In [9]:
#importing ords_prods_merge dataframe
df_ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merge.pkl'))

In [10]:
df_ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   3139998      138            28                   6                 11   
1   1977647      138            30                   6                 17   
2    389851      709             2                   0                 21   
3    652770      764             1                   3                 13   
4   1813452      764             3                   4                 17   

   days_since_last_order  first_order  add_to_cart_order  reordered _merge  
0                    3.0            0                  5          0   both  
1                   20.0            0                  1          1   both  
2                    6.0            0                 20          0   both  
3                    NaN            1                 10          0   both  
4                    9.0            0                 11          1   both

In [11]:
#creating a subset of the first 1000000 rows 
df = df_ords_prods_merge[:1000000]

In [12]:
#checking the shape of the new subset 
df.shape

(1000000, 15)

# 03. Creating a "price_label" column 

In [14]:
#applying a string label to every row within the dataframe, designating it as a low-, mid-, or high-range product based on its price.
def price_label(row):

  if row['prices'] <= 5:
    return 'Low-range product'
  elif (row['prices'] > 5) and (row['prices'] <= 15):
    return 'Mid-range product'
  elif row['prices'] > 15:
    return 'High range'
  else: return 'Not enough data'

In [15]:
#applying price_label function to subset
df['price_range'] = df.apply(price_label, axis=1)

C:\Users\canam\AppData\Local\Temp\ipykernel_15952\1800188404.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_range'] = df.apply(price_label, axis=1)


In [16]:
#checking values of the new column 
df['price_range'].value_counts(dropna=False)

price_range
Mid-range product    652638
Low-range product    338018
High range             9344
Name: count, dtype: int64

In [17]:
#checking the most expensive item in subset
df['prices'].max()

24.5

# 04. Creating if statements with the loc() function

In [64]:
#If statements with the loc() function
df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [20]:
df.loc[(df['prices'] <= 15) & (df['prices'] > 5), 'price_range_loc'] = 'Mid-range product'

In [21]:
df.loc[df['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [22]:
#counting the number of values within each label
df['price_range_loc'].value_counts(dropna=False)

price_range_loc
Mid-range product     652638
Low-range product     338018
High-range product      9344
Name: count, dtype: int64

In [23]:
#Repeating the process with entire ords_prods_merge dataframe
df_ords_prods_merge.loc[df_ords_prods_merge['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [24]:
df_ords_prods_merge.loc[(df_ords_prods_merge['prices'] <= 15) & (df_ords_prods_merge['prices'] > 5), 'price_range_loc'] = 'Mid-range product'

In [25]:
df_ords_prods_merge.loc[df_ords_prods_merge['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [26]:
df_ords_prods_merge['price_range_loc'].value_counts(dropna=False)

price_range_loc
Mid-range product     21865987
Low-range product     10126321
High-range product      412551
Name: count, dtype: int64

# 05.  Creating a "busiest_day" column

In [28]:
#determining when orders take place
df_ords_prods_merge['orders_day_of_week'].value_counts(dropna=False)

orders_day_of_week
0    6204182
1    5660230
6    4496490
2    4213830
5    4205791
3    3840534
4    3783802
Name: count, dtype: int64

In [29]:
#creating a new column, “busiest_day,” that will contain one of three different values: “Busiest day,” “Least busy,” and “Regularly busy.” using for-loop
result = []

for value in df_ords_prods_merge["orders_day_of_week"]:
  if value == 0:
    result.append("Busiest day")
  elif value == 4:
    result.append("Least busy")
  else:
    result.append("Regularly busy")

In [30]:
#printing the result 
result

['Regularly busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Busiest day',
 'Busiest day',
 'Busiest day',
 'Busiest day',
 'Busiest day',
 'Least busy',
 'Regularly busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Regularly busy',
 'Least busy',
 'Regularly busy',
 'Busiest day',
 'Regularly busy',
 'Busiest day',
 'Least busy',
 'Busiest day',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Regularly busy',
 'Bus

In [31]:
#combining results with the ords_prods_merge dataframe 
df_ords_prods_merge['busiest_day']=result

In [32]:
#counting the values 
df_ords_prods_merge['busiest_day'].value_counts(dropna = False)

busiest_day
Regularly busy    22416875
Busiest day        6204182
Least busy         3783802
Name: count, dtype: int64

# 06. Creating new "busiest days" column

Suppose your clients have changed their minds about the labels you created in your “busiest_day” column. Now, they want “Busiest day” to become “Busiest days” (plural). This label should correspond with the two busiest days of the week as opposed to the single busiest day. At the same time, they’d also like to know the two slowest days. Create a new column for this using a suitable method.

In [34]:
#creating a new column, “busiest days,” that will contain one of three different values: “Busiest days,” “Slowest Days,” and “Regularly busy.”
#Busiest days are day 0 and 1
#Slowest days are day 4 and 3 
result_2 = []

for value in df_ords_prods_merge["orders_day_of_week"]:
  if value == 0 or value == 1:
    result_2.append("Busiest days")
  elif value == 4 or value == 3:
    result_2.append("Slowest days")
  else:
    result_2.append("Regularly busy")

In [35]:
# combining results with the ords_prods_merge dataframe 
df_ords_prods_merge['busiest_days']=result_2

In [36]:
#checking the values
df_ords_prods_merge['busiest_days'].value_counts(dropna = False)

busiest_days
Regularly busy    12916111
Busiest days      11864412
Slowest days       7624336
Name: count, dtype: int64

In [37]:
df_ords_prods_merge.shape

(32404859, 18)

The total counts for regularly busy, busiest days, and slowest days add up to 32404859, which is the number of rows in the df_ords_prods_merge data frame.  In addition, the busiest days count is equal to the sum of orders_day_of_week 0 and orders_day_of_week 1. Similarly, the slowest days count is equal to the sum of orders_day_of_week 4 and orders_day_of_week 3.  

In [39]:
#checking to see if busiest_days column has been added 
df_ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   3139998      138            28                   6                 11   
1   1977647      138            30                   6                 17   
2    389851      709             2                   0                 21   
3    652770      764             1                   3                 13   
4   1813452      764             3                   4                 17   

   days_since_last_order  first_order  add_to_cart_order  reordered _merge  \
0                    3.0            0                  5          0   both   
1                   20.0            0                  1          1   both   
2                    6.0            0                 20          0   both   
3                    NaN            1                 10          0   both   
4                    9.0            0                 11          1   both   

     price_range_loc     busiest_day    busiest_days  
0  Mid-range product  Regularly busy  Regularly busy  
1  Mid-range product  Regularly busy  Regularly busy  
2  Mid-range product     Busiest day    Busiest days  
3  Mid-range product  Regularly busy    Slowest days  
4  Mid-range product      Least busy    Slowest days

# 07. Creating "busiest period of day" column

When too many users make Instacart orders at the same time, the app freezes. The senior technical officer at Instacart wants you to identify the busiest hours of the day. Rather than by hour, they want periods of time labeled “Most orders,” “Average orders,” and “Fewest orders.” Create a new column containing these labels called “busiest_period_of_day.”

In [40]:
#creating a new column busiest_period_of_day that will contain lables “Most orders,” “Average orders,” and “Fewest orders.” 
#First, we must check the value counts in the order_hour_of_day column.
df_ords_prods_merge['order_hour_of_day'].value_counts(dropna = False)

order_hour_of_day
10    2761760
11    2736140
14    2689136
15    2662144
13    2660954
12    2618532
16    2535202
9     2454203
17    2087654
8     1718118
18    1636502
19    1258305
20     976156
7      891054
21     795637
22     634225
23     402316
6      290493
0      218769
1      115700
5       87961
2       69375
4       53242
3       51281
Name: count, dtype: int64

We must divide order_hour_of_day into 3 sections.  
Most orders: 10, 11, 14, 15, 13, 12, 16, 9
Fewest orders: 23, 6, 0, 1, 5, 2, 4, 3
Average orders: All other values 

In [82]:
#creating a for-loop to label periods of time   
result_3 = []

for value in df_ords_prods_merge["order_hour_of_day"]:
  if value in [10, 11, 14, 15, 13, 12, 16, 9]:
    result_3.append("Most Orders")
  elif value in [23, 6, 0, 1, 5, 2, 4, 3]:
    result_3.append("Fewest Orders")
  else:
    result_3.append("Average Orders")

In [83]:
# combining results with the ords_prods_merge dataframe 
df_ords_prods_merge['busiest_period_of_day']=result_3

In [84]:
#checking the frequency for the new column
df_ords_prods_merge['busiest_period_of_day'].value_counts(dropna = False)

busiest_period_of_day
Most Orders       21118071
Average Orders     9997651
Fewest Orders      1289137
Name: count, dtype: int64

In [80]:
#checking to see that busiest_period_of_day column has been added
df_ords_prods_merge.head()

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2           1  Chocolate Sandwich Cookies        61             19     5.8   
3           1  Chocolate Sandwich Cookies        61             19     5.8   
4           1  Chocolate Sandwich Cookies        61             19     5.8   

   order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   3139998      138            28                   6                 11   
1   1977647      138            30                   6                 17   
2    389851      709             2                   0                 21   
3    652770      764             1                   3                 13   
4   1813452      764             3                   4                 17   

   days_since_last_order  first_order  add_to_cart_order  reordered _merge  \
0                    3.0            0                  5          0   both   
1                   20.0            0                  1          1   both   
2                    6.0            0                 20          0   both   
3                    NaN            1                 10          0   both   
4                    9.0            0                 11          1   both   

     price_range_loc     busiest_day    busiest_days busiest_period_of_day  
0  Mid-range product  Regularly busy  Regularly busy           Most Orders  
1  Mid-range product  Regularly busy  Regularly busy        Average Orders  
2  Mid-range product     Busiest day    Busiest days        Average Orders  
3  Mid-range product  Regularly busy    Slowest days           Most Orders  
4  Mid-range product      Least busy    Slowest days        Average Orders

# 08. Exporting Data 

In [88]:
df_ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'ords_prods_merge_derived.pkl'))